python_modules/dagster/dagster/_core/definitions/asset_checks.py

```diff
@@ -127,3 +141,46 @@ def get_attributes_dict(self) -> Dict[str, Any]:
             specs=self._specs,
             input_output_props=self._input_output_props,
         )
+
+
+@experimental
+def build_blocking_asset_check(
+    asset: "AssetsDefinition",
+    checks: Sequence[AssetChecksDefinition],
+):
+    from dagster import In, graph_asset, op, Output
+
+    # key_prefix=asset.key[0:-1] if len(asset.key) > 1 else None,
+    asset_key = "blocking_" + asset.key.path[-1]
+
+    check_specs = []
+    for c in checks:
+        check_specs.extend(c.specs)
+
+    check_output_names = [c.get_python_identifier() for c in check_specs]
+
+    @op(ins={"materialization": In(Any), "check_evaluations": In(Any)})
+    def fan_in_checks_and_materialization(context, materialization, check_evaluations):
+        yield Output(materialization)
+
+        for result in check_evaluations:
+            if not result.success:
+                raise Exception("Check failed")
```

Comment:
Raise a specific exception type (should definitely not throw bare exceptions in any cases) and then we can put custom metadata on it and potentially render it specially.

At a minimum you should construct a much more informative error string with all the context a user would want